File: src/Monet/MOEvent.m
#import <Monet/MOEvent.h>

struct MOEventData
{
	MOEventType     type;
	uint8_t           modifiers;
    
	NSString        *character;
	MOKey           key;
    
	MOMouseButton   mouseButton;
	MOPoint         mouseLocation;
	MOPoint         relativeMouseMotion;
	uint8_t           clickCount;
};

@implementation MOEvent

- (id)initKeyEventWithType: (MOEventType)aType modifiers: (uint8_t)aModifiers character: (NSString *)aCharacter key: (MOKey)aKey
{
	if (self = [super init])
	{
		eventData = calloc(1, sizeof (struct MOEventData));

		eventData->type      = aType;
		eventData->modifiers = aModifiers;

		eventData->character = [aCharacter retain];
		eventData->key       = aKey;
	}

	return self;
}

- (id)initMouseButtonEventWithType: (MOEventType)aType modifiers: (uint8_t)aModifiers mouseButton: (MOMouseButton)aMouseButton mouseLocation: (MOPoint)aMouseLocation clickCount: (uint8_t)aClickCount
{
	if (self = [super init])
	{
		eventData = calloc(1, sizeof (struct MOEventData));

		eventData->type          = aType;
		eventData->modifiers     = aModifiers;
        
		eventData->mouseButton   = aMouseButton;
		eventData->mouseLocation = aMouseLocation;
		eventData->clickCount    = aClickCount;
	}

	return self;
}

- (id)initMouseMotionEventWithModifiers: (uint8_t)aModifiers mouseLocation: (MOPoint)aMouseLocation relativeMouseMotion: (MOPoint)aRelativeMouseMotion
{
	if (self = [super init])
	{
		eventData = calloc(1, sizeof (struct MOEventData));

		eventData->type                = MOMouseMotionEventType;
		eventData->modifiers           = aModifiers;
        
		eventData->mouseLocation       = aMouseLocation;
		eventData->relativeMouseMotion = aRelativeMouseMotion;
	}

	return self;
}

- (void)dealloc
{
	[eventData->character release];

	[super dealloc];
}

#pragma mark -

- (MOEventType)type
{
	return eventData->type;
}

- (uint8_t)modifiers
{
	return eventData->modifiers;
}

#pragma mark -

- (NSString *)character
{
	return eventData->character;
}

- (MOKey)key
{
	return eventData->key;
}

#pragma mark -

- (MOMouseButton)mouseButton
{
	return eventData->mouseButton;
}

- (MOPoint)mouseLocation
{
	return eventData->mouseLocation;
}

- (MOPoint)relativeMouseMotion
{
	return eventData->relativeMouseMotion;
}

- (uint8_t)clickCount
{
	return eventData->clickCount;
}

@end

File: src/Monet/MOEvent.c
#include <Monet/MOEvent.h>

#include <string.h>

#include <Monet/Common.h>

struct _MOEvent
{
	COGuts          *guts;

	MOEventType     type;
	uint8_t         modifiers;

	char            *character;
	MOKey           key;

	MOMouseButton   mouseButton;
	MOPoint         mouseLocation;
	MOPoint         relativeMouseMotion;
	uint8_t         clickCount;

	MOTimer         *timer;
};

void _MOEventDestroy(void *aEvent);

MOEvent *MOEventCreateKey(MOEventType aType, MOKeyModifierMask aModifiers, char *aCharacter, MOKey aKey)
{
	MOEvent *event = calloc(1, sizeof (MOEvent));
	COInitialize(event);
	COSetDestructor(event, &_MOEventDestroy);

	event->type      = aType;
	event->modifiers = aModifiers;

	event->character = aCharacter ? strdup(aCharacter) : NULL;
	event->key       = aKey;

	return event;
}

MOEvent *MOEventCreateMouseButton(MOEventType aType, MOKeyModifierMask aModifiers, MOMouseButton aMouseButton, MOPoint aMouseLocation, uint8_t aClickCount)
{
	MOEvent *event = calloc(1, sizeof (MOEvent));
	COInitialize(event);
	COSetDestructor(event, &_MOEventDestroy);

	event->type          = aType;
	event->modifiers     = aModifiers;

	event->mouseButton   = aMouseButton;
	event->mouseLocation = aMouseLocation;
	event->clickCount    = aClickCount;

	return event;
}

MOEvent *MOEventCreateMouseMotion(MOKeyModifierMask aModifiers, MOPoint aMouseLocation, MOPoint aRelativeMouseMotion)
{
	MOEvent *event = calloc(1, sizeof (MOEvent));
	COInitialize(event);
	COSetDestructor(event, &_MOEventDestroy);

	event->type                = MOMouseMotionEventType;
	event->modifiers           = aModifiers;
        
	event->mouseLocation       = aMouseLocation;
	event->relativeMouseMotion = aRelativeMouseMotion;

	return event;
}

MOEvent *MOEventCreateTimer(MOTimer *aTimer)
{
	MOEvent *event = calloc(1, sizeof (MOEvent));
	COInitialize(event);
	COSetDestructor(event, &_MOEventDestroy);

	event->type  = MOTimerFiredEventType;

	event->timer = CORetain(aTimer);

	return event;
}

void _MOEventDestroy(void *aEvent)
{
	MOEvent *event = (MOEvent *)aEvent;

	if (event->character)
		free(event->character);
	CORelease(event->timer);
}

MOEventType MOEventGetType(MOEvent *aEvent)
{
	return aEvent->type;
}

MOKeyModifierMask MOEventGetModifiers(MOEvent *aEvent)
{
	return aEvent->modifiers;
}

char *MOEventGetCharacter(MOEvent *aEvent)
{
	return aEvent->character;
}

MOKey MOEventGetKey(MOEvent *aEvent)
{
	return aEvent->key;
}

MOMouseButton MOEventGetMouseButton(MOEvent *aEvent)
{
	return aEvent->mouseButton;
}

MOPoint MOEventGetMouseLocation(MOEvent *aEvent)
{
	return aEvent->mouseLocation;
}

MOPoint MOEventGetRelativeMouseMotion(MOEvent *aEvent)
{
	return aEvent->relativeMouseMotion;
}

uint8_t MOEventGetClickCount(MOEvent *aEvent)
{
	return aEvent->clickCount;
}

MOTimer *MOEventGetTimer(MOEvent *aEvent)
{
	return aEvent->timer;
}
